File: services/notification-service/src/test/java/com/visor/school/notification/service/NotificationServiceTest.java
package com.visor.school.notification.service;

import com.visor.school.notification.model.Notification;
import com.visor.school.notification.model.NotificationType;
import com.visor.school.notification.repository.NotificationRepository;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
class NotificationServiceTest {

    @Mock
    private NotificationRepository notificationRepository;

    @Mock
    private NotificationDeliveryService notificationDeliveryService;

    private NotificationService notificationService;

    private final UUID testUserId = UUID.randomUUID();

    @BeforeEach
    void setup() {
        notificationService = new NotificationService(notificationRepository, notificationDeliveryService);
    }

    @Test
    void shouldCreateNotification() {
        // Given
        Notification notification = new Notification(
            testUserId,
            NotificationType.GRADE_POSTED,
            "New Grade Posted",
            "A new grade has been posted",
            null
        );
        when(notificationRepository.save(any(Notification.class))).thenReturn(notification);
        doNothing().when(notificationDeliveryService).deliver(any(Notification.class));

        // When
        Notification result = notificationService.create(
            testUserId,
            NotificationType.GRADE_POSTED,
            "New Grade Posted",
            "A new grade has been posted",
            null
        );

        // Then
        assertNotNull(result);
        assertEquals(testUserId, result.getUserId());
        assertEquals(NotificationType.GRADE_POSTED, result.getType());
        verify(notificationRepository).save(any(Notification.class));
        verify(notificationDeliveryService).deliver(any(Notification.class));
    }

    @Test
    void shouldMarkNotificationAsRead() {
        // Given
        UUID notificationId = UUID.randomUUID();
        Notification notification = new Notification(
            testUserId,
            NotificationType.GRADE_POSTED,
            "New Grade",
            "Your grade has been posted",
            null
        );
        notification.setId(notificationId);
        notification.setRead(false);

        Notification readNotification = new Notification(
            testUserId,
            NotificationType.GRADE_POSTED,
            "New Grade",
            "Your grade has been posted",
            null
        );
        readNotification.setId(notificationId);
        readNotification.setRead(true);

        when(notificationRepository.findById(notificationId)).thenReturn(Optional.of(notification));
        when(notificationRepository.save(any(Notification.class))).thenReturn(readNotification);

        // When
        Notification result = notificationService.markAsRead(notificationId);

        // Then
        assertTrue(result.isRead());
        verify(notificationRepository).findById(notificationId);
        verify(notificationRepository).save(any(Notification.class));
    }

    @Test
    void shouldGetNotificationsByUser() {
        // Given
        List<Notification> notifications = List.of(
            new Notification(
                testUserId,
                NotificationType.GRADE_POSTED,
                "Grade 1",
                "Message 1",
                null
            ),
            new Notification(
                testUserId,
                NotificationType.ATTENDANCE_MARKED,
                "Attendance 1",
                "Message 2",
                null
            )
        );
        when(notificationRepository.findByUserIdOrderByCreatedAtDesc(testUserId)).thenReturn(notifications);

        // When
        List<Notification> result = notificationService.getByUser(testUserId);

        // Then
        assertEquals(2, result.size());
        verify(notificationRepository).findByUserIdOrderByCreatedAtDesc(testUserId);
    }

    @Test
    void shouldGetUnreadNotificationsByUser() {
        // Given
        List<Notification> notifications = List.of(
            new Notification(
                testUserId,
                NotificationType.GRADE_POSTED,
                "Grade 1",
                "Message 1",
                null
            )
        );
        when(notificationRepository.findByUserIdAndReadFalseOrderByCreatedAtDesc(testUserId)).thenReturn(notifications);

        // When
        List<Notification> result = notificationService.getUnread(testUserId);

        // Then
        assertEquals(1, result.size());
        verify(notificationRepository).findByUserIdAndReadFalseOrderByCreatedAtDesc(testUserId);
    }
}

File: services/notification-service/src/test/kotlin/com/visor/school/notification/service/NotificationServiceTest.kt
package com.visor.school.notification.service

import com.visor.school.notification.model.Notification
import com.visor.school.notification.model.NotificationType
import com.visor.school.notification.repository.NotificationRepository
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.junit.jupiter.MockitoExtension
import org.mockito.kotlin.*
import java.util.UUID

@ExtendWith(MockitoExtension::class)
class NotificationServiceTest {

    @Mock
    private lateinit var notificationRepository: NotificationRepository

    @Mock
    private lateinit var notificationDeliveryService: NotificationDeliveryService

    @InjectMocks
    private lateinit var notificationService: NotificationService

    private val testUserId = UUID.randomUUID()

    @BeforeEach
    fun setup() {
        notificationService = NotificationService(notificationRepository, notificationDeliveryService)
    }

    @Test
    fun `should create notification`() {
        // Given
        val notification = Notification(
            userId = testUserId,
            type = NotificationType.GRADE_POSTED,
            title = "New Grade Posted",
            message = "A new grade has been posted"
        )
        whenever(notificationRepository.save(any())).thenReturn(notification)
        doNothing().whenever(notificationDeliveryService).deliver(any())

        // When
        val result = notificationService.create(
            userId = testUserId,
            type = NotificationType.GRADE_POSTED,
            title = "New Grade Posted",
            message = "A new grade has been posted"
        )

        // Then
        assertNotNull(result)
        assertEquals(testUserId, result.userId)
        assertEquals(NotificationType.GRADE_POSTED, result.type)
        verify(notificationRepository).save(any())
        verify(notificationDeliveryService).deliver(any())
    }

    @Test
    fun `should mark notification as read`() {
        // Given
        val notificationId = UUID.randomUUID()
        val notification = Notification(
            id = notificationId,
            userId = testUserId,
            type = NotificationType.GRADE_POSTED,
            title = "New Grade",
            message = "Your grade has been posted",
            read = false
        )
        whenever(notificationRepository.findById(notificationId)).thenReturn(java.util.Optional.of(notification))
        whenever(notificationRepository.save(any())).thenReturn(notification.copy(read = true))

        // When
        val result = notificationService.markAsRead(notificationId)

        // Then
        assertTrue(result.read)
        verify(notificationRepository).findById(notificationId)
        verify(notificationRepository).save(any())
    }

    @Test
    fun `should get notifications by user`() {
        // Given
        val notifications = listOf(
            Notification(
                userId = testUserId,
                type = NotificationType.GRADE_POSTED,
                title = "Grade 1",
                message = "Message 1"
            ),
            Notification(
                userId = testUserId,
                type = NotificationType.ATTENDANCE_MARKED,
                title = "Attendance 1",
                message = "Message 2"
            )
        )
        whenever(notificationRepository.findByUserIdOrderByCreatedAtDesc(testUserId)).thenReturn(notifications)

        // When
        val result = notificationService.getByUser(testUserId)

        // Then
        assertEquals(2, result.size)
        verify(notificationRepository).findByUserIdOrderByCreatedAtDesc(testUserId)
    }

    @Test
    fun `should get unread notifications`() {
        // Given
        val notifications = listOf(
            Notification(
                userId = testUserId,
                type = NotificationType.GRADE_POSTED,
                title = "Unread 1",
                message = "Message 1",
                read = false
            ),
            Notification(
                userId = testUserId,
                type = NotificationType.ATTENDANCE_MARKED,
                title = "Unread 2",
                message = "Message 2",
                read = false
            )
        )
        whenever(notificationRepository.findByUserIdAndReadFalseOrderByCreatedAtDesc(testUserId)).thenReturn(notifications)

        // When
        val result = notificationService.getUnread(testUserId)

        // Then
        assertEquals(2, result.size)
        assertTrue(result.all { !it.read })
        verify(notificationRepository).findByUserIdAndReadFalseOrderByCreatedAtDesc(testUserId)
    }
}

